### ISSUE TITLE
Error when python lib in multiple python version is in PATH

### ISSUE THREAD

**Issue by davidak** [Mic92/nixpkgs-review#194]
This PR adds 2 python libs that are available for 2 python versions.

https://github.com/NixOS/nixpkgs/pull/120453

5 packages built:

- python38Packages.humblewx
- python38Packages.pysvg-py3
- python39Packages.humblewx
- python39Packages.pysvg-py3
- timeline

When nixpkgs-review puts me in the nix-shell, the main program don't work because it uses the wrong lib.

https://github.com/NixOS/nixpkgs/pull/120453#issuecomment-826063089

When i specify just the main package, i can run it:

`nix run nixpkgs-review -c nixpkgs-review pr 120453 -p timeline`

So i think this is a problem with nixpkgs-review.

We should improve the UX.

At least show a warning in such cases that it might lead to issues and suggest `-p`.

Or test in separate nix-shells?

**Comment by SuperSandro2000** [comment]
Most likely this is broken because we unset PYTHONPATH https://github.com/Mic92/nixpkgs-review/blob/master/default.nix#L36

**Comment by Mic92** [comment]
No. It's not because we unset PYTHONPATH but because multiple python versions are mixed. 
I don't know a good solution to this. It has been also reported in the past here:
 https://github.com/Mic92/nixpkgs-review/issues/46

**Comment by Atemu** [comment]
Actually, we don't unset it. It's still in the env when in the shell and that's also what causes applications to fail.

Unsetting it makes everything work fine.

**Comment by davidak** [comment]
This is still an issue: https://github.com/NixOS/nixpkgs/pull/221908#issuecomment-1475374419

**Comment by Mic92** [comment]
> Unsetting it makes everything work fine.

But than testing python libraries becomes harder as well.

**Comment by Mic92** [comment]
Do those python applications not ignore PYTHONPATH or was it just that they do not propagated buildInputs?

**Comment by doronbehar** [comment]
I think I experienced today this issue in https://github.com/NixOS/nixpkgs/pull/269358#pullrequestreview-1767600855 where several Python modules were added, and a Python (GUI) application was added as well. In the `$PYTHONPATH` I saw only Python3.10 packages, and the main GUI application didn't launch because it was built using Python3.11. I'm not sure if this is the same issue as reported here, because only Python3.10 packages were found at `$PYTHONPATH` in `nixpkgs-review`'s shell, and I don't remember I ever experienced this issue before.

I also don't understand why is `$PYTHONPATH` set in the first place. If I want to test new Python libraries in a review, perhaps that should be part of the expression - why forcibly inject a value to `$PYTHONPATH`?

**Comment by Mic92** [comment]
It's not forced, it's was nix-shell does by default. We would need to explicitly unset it.

**Comment by Atemu** [comment]
Oh, right, we land inside a `nix-shell`. Would it be feasible to land inside a `nix shell` instead?

**Comment by doronbehar** [comment]
I tested the following:

```
[ original shell ] $ nix-shell -p python310Packages.numpy -p python311Packages.numpy
(ins)[nix-shell:~]$ echo $PYTHONPATH | sed 's/:/\n/g'
/nix/store/flcg981mnk58jiwnf2cfa64z11j37fyp-python3.11-numpy-1.26.1/lib/python3.11/site-packages
/nix/store/qp5zys77biz7imbk6yy85q5pdv7qk84j-python3-3.11.6/lib/python3.11/site-packages
```

V.S:
```
[ original shell ] $ nix-shell -p python311Packages.numpy -p python310Packages.numpy
(cmd)[nix-shell:~]$ echo $PYTHONPATH | sed 's/:/\n/g'
/nix/store/kcx6gnm85y267954i0a05r90g4n0cahd-python3.10-numpy-1.26.1/lib/python3.10/site-packages
/nix/store/50kabpj0s79040a42b7jj2dxn85wmbfd-python3-3.10.13/lib/python3.10/site-packages
```

Meaning that the later version is ruling the shell. The `nix shell` behavior OTH:

```
[ original shell ] $ nix shell nixpkgs#python310Packages.numpy nixpkgs#python311Packages.numpy
[ looks like original shell ] $ echo $PYTHONPATH

```

I think we all agree that the later behavior is more desirable. Could it be the only drawback is that it requires using Nix' 'experimental' features?

**Comment by Mic92** [comment]
`nix shell nixpkgs#python310Packages.numpy nixpkgs#python311Packages.numpy` will go out memory for a larger set of packages easily because it creates a nixpkgs instance for each package. However you `buildEnv` could be used instead.

**Comment by tobiasBora** [comment]
Note that it would be cool if a solution to this thread also gives a solution to this thread https://github.com/Mic92/nixpkgs-review/issues/484#issuecomment-2809025464

Basically, in the current thread PYTHONPATH must be unset for things to work (because we run a program that already uses `python.withPackages` internally I guess), while in the above issue, PYTHONPATH is needed when testing a python library, but if the library is available for multiple python versions, the path only contains ONE version, but `python` points to the other versions, hence crashing. A solution that would work in both cases would be to unset `PYTHONPATH` + use `pythonX.withPackages` (`X` spanning all versions available for the library), but it might not be so simple to implement.
